src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,351 @@
+      lms_gen_leaf(out, lmots_pk, tree_address, *hash);
+   };
+}
+
+void lms_treehash(StrongSpan<LMS_Tree_Node> out_root,
+                  std::optional<StrongSpan<LMS_AuthenticationPath>> out_auth_path,
+                  std::optional<LMS_Tree_Node_Idx> leaf_idx,
+                  const LMS_PrivateKey& lms_sk) {
+   auto hash_pair_func = get_hash_pair_func_for_identifier(lms_sk.lms_params(), lms_sk.identifier());
+   auto gen_leaf = lms_gen_leaf_func(lms_sk);
+   TreeAddress lms_tree_address(lms_sk.lms_params().h());
+
+   treehash(out_root,
+            out_auth_path,
+            leaf_idx,
+            lms_sk.lms_params().m(),
+            LMS_TreeLayerIndex(lms_sk.lms_params().h()),
+            0,
+            hash_pair_func,
+            gen_leaf,
+            lms_tree_address);
+}
+
+}  // namespace
+
+LMS_Params LMS_Params::create_or_throw(LMS_Algorithm_Type type) {
+   uint8_t type_value = checked_cast_to_or_throw<uint8_t, Decoding_Error>(type, "Unsupported LMS algorithm type");
+
+   if(type >= LMS_Algorithm_Type::SHA256_M32_H5 && type <= LMS_Algorithm_Type::SHA256_M32_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHA256_M32_H5) + 1);
+      return LMS_Params(type, "SHA-256", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHA256_M24_H5 && type <= LMS_Algorithm_Type::SHA256_M24_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHA256_M24_H5) + 1);
+      return LMS_Params(type, "Truncated(SHA-256,192)", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHAKE_M32_H5 && type <= LMS_Algorithm_Type::SHAKE_M32_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHAKE_M32_H5) + 1);
+      return LMS_Params(type, "SHAKE-256(256)", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHAKE_M24_H5 && type <= LMS_Algorithm_Type::SHAKE_M24_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHAKE_M24_H5) + 1);
+      return LMS_Params(type, "SHAKE-256(192)", h);
+   }
+
+   throw Decoding_Error("Unsupported LMS algorithm type");
+}
+
+LMS_Params LMS_Params::create_or_throw(std::string_view hash_name, uint8_t h) {
+   BOTAN_ARG_CHECK(h == 5 || h == 10 || h == 15 || h == 20 || h == 25, "Invalid h value");
+   auto type_offset = h / 5 - 1;
+   LMS_Algorithm_Type base_type;
+
+   if(hash_name == "SHA-256") {
+      base_type = LMS_Algorithm_Type::SHA256_M32_H5;
+   } else if(hash_name == "Truncated(SHA-256,192)") {
+      base_type = LMS_Algorithm_Type::SHA256_M24_H5;
+   } else if(hash_name == "SHAKE-256(256)") {
+      base_type = LMS_Algorithm_Type::SHAKE_M32_H5;
+   } else if(hash_name == "SHAKE-256(192)") {
+      base_type = LMS_Algorithm_Type::SHAKE_M24_H5;
+   } else {
+      throw Decoding_Error("Unsupported hash function");
+   }
+   auto type = checked_cast_to<LMS_Algorithm_Type>(checked_cast_to<uint8_t>(base_type) + type_offset);
+   return LMS_Params(type, hash_name, h);
+}
```

Comment:
Same as above: enum-arithmetics, really? It's beautiful though, I'll admit that.
Please, leave a big and blinking warning in the `enum` declaration that you're doing this, at least. Future-you will be grateful.

Reply:
I'll also use switches for lm_ots, where we did the same 👍